hail/python/hail/ir/blockmatrix_ir.py

```diff
@@ -19,21 +20,76 @@ def _compute_type(self):
         self._type = Env.backend().blockmatrix_type(self)
 
 
-class BlockMatrixAdd(BlockMatrixIR):
-    @typecheck_method(left=BlockMatrixIR, right=BlockMatrixIR)
-    def __init__(self, left, right):
+class BlockMatrixMap2(BlockMatrixIR):
+    @typecheck_method(left=BlockMatrixIR, right=BlockMatrixIR, apply_bin_op=ApplyBinaryOp)
+    def __init__(self, left, right, apply_bin_op):
         super().__init__()
         self._left = left
         self._right = right
+        self._apply_bin_op = apply_bin_op
 
     def render(self, r):
-        return f'(BlockMatrixAdd {r(self._left)} {r(self._right)})'
+        return f'(BlockMatrixMap2 {r(self._left)} {r(self._right)} {r(self._apply_bin_op)})'
 
     def _compute_type(self):
-        self._right.typ # force
+        self._right.typ  # Force
```

Comment:
What's the motivation for this?

Reply:
I also think this is weird.

Reply:
@danking So while computing types on construction could be done for non-value IRs, IR nodes need an environment which may not be known eagerly. This is why children are type checked in `_compute_type()`. Seems like not having eager checks in non-value IRs is mainly for consistency. Could either make eager type checks in non-value IRs or have a function like `check_type()` that would cache, but I think that's kind of confusing given the current setup of the `typ` property already caching.